### PR TITLE
Fix bug with incorrect package name fixing

### LINKED ISSUE #875: Package name fixing incorrect with file-level directives
Package name rule incorrectly adds package directive to the following code:

```
@file:Suppress("CONSTANT_UPPERCASE")

// FixMe: we need to use package name here when https://github.com/gradle/gradle/issues/9270 will be fixed
// now it is just a workaround

object Versions {
    const val KOTLIN = "1.4.10"
    const val JUNIT = "5.7.1"
    const val OKIO = "3.0.0-alpha.1"
    const val SERIALIZATION = "1.1.0"
}

```

To reproduce use:
` ./ktlint -R diktat-0.5.2.jar --disabled_rules=standard -F --debug "Test.kt"`

### PR DESCRIPTION
**Pull request by kgevorkyan**
### What's done:
* Fix bug with incorrect package name fixing when there is exist file annotation
* Add tests

Closes #875

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt [hunk 1]
```diff
@@ -219,7 +219,12 @@ class PackageNaming(configRules: List<RulesConfig>) : DiktatRule(
                 // there is missing package statement in a file, so it will be created and inserted
                 val newPackageDirective = generatedPackageDirective.findLeafWithSpecificType(PACKAGE_DIRECTIVE)!!
                 packageDirectiveNode.treeParent.replaceChild(packageDirectiveNode, newPackageDirective)
-                newPackageDirective.treeParent.addChild(PsiWhiteSpaceImpl("\n\n"), newPackageDirective.treeNext)
+                val insertPackageDirectiveBefore = if (newPackageDirective != newPackageDirective.treeParent.firstChildNode) {
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:222]: make i with ternary operator

**Review by akuleshov7 — APPROVED** [review]
lgtm

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt [hunk 2]
```diff
@@ -219,7 +219,12 @@ class PackageNaming(configRules: List<RulesConfig>) : DiktatRule(
...
+                    newPackageDirective
+                } else {
+                    newPackageDirective.treeNext
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:226]: What if there will be `EOL_COMMENT` or another type of comment? I suppose, comment will be in wrong line.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt [hunk 3]
```diff
@@ -218,8 +218,14 @@ class PackageNaming(configRules: List<RulesConfig>) : DiktatRule(
             ?: run {
                 // there is missing package statement in a file, so it will be created and inserted
                 val newPackageDirective = generatedPackageDirective.findLeafWithSpecificType(PACKAGE_DIRECTIVE)!!
-                packageDirectiveNode.treeParent.replaceChild(packageDirectiveNode, newPackageDirective)
-                newPackageDirective.treeParent.addChild(PsiWhiteSpaceImpl("\n\n"), newPackageDirective.treeNext)
+                val packageDirectiveParent = packageDirectiveNode.treeParent
+                packageDirectiveParent.replaceChild(packageDirectiveNode, newPackageDirective)
+                val insertWhitespaceBefore = if (newPackageDirective != packageDirectiveParent.firstChildNode) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:223]: Maybe, check whether there is a white space node before and after package directive node? Actually, these `\n\n` are not responsibility of this rule, so here we should ensure that the tree is correct, and FileStructureRule will do the rest.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt (hunk 2 above)
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:226]: Denis, very good note. thanks!

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt (hunk 3 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:223]: But shouldn't each rule work independently? What if user disable `FileStructureRule` ?

Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:223]: Yes, rules should work independently. My main suggestion was to check, whether there are white spaces both before and after the package directive node; I think we can make two newlines on both sides.

Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:223]: Got it. Yes, I agree

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt [hunk 4]
```diff
@@ -218,8 +224,27 @@ class PackageNaming(configRules: List<RulesConfig>) : DiktatRule(
...
+                val packageDirectiveParent = packageDirectiveNode.treeParent
+                // When package directive is missing in .kt file,
+                // the node is still present in the AST, and not always in a convenient place.
+                // E.g. `@file:Suppress("...") // comments`
+                // AST will be: FILE_ANNOTATION_LIST, PACKAGE_DIRECTIVE, WHITE_SPACE, EOL_COMMENT
+                // So, we can't just put new package directive in it's old place and rely on FileStructure rule
+                if (packageDirectiveNode != packageDirectiveParent.firstChildNode) {
+                    // We will insert new package directive node before first node, which is not in the following list
+                    val possibleTypesBeforePackageDirective = listOf(WHITE_SPACE, EOL_COMMENT, BLOCK_COMMENT, KDOC, PACKAGE_DIRECTIVE, FILE_ANNOTATION_LIST)
+                    val addBefore = packageDirectiveParent.children().filter { it.elementType !in possibleTypesBeforePackageDirective }.first()
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:236]:
```suggestion
                    val addBefore = packageDirectiveParent.children().first { it.elementType !in possibleTypesBeforePackageDirective }
```

Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt:236]:
```suggestion
                    val addBefore = packageDirectiveParent.children().first { it.elementType !in possibleTypesBeforePackageDirective }
```

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/TopLevelOrderRuleFixTest.kt [hunk 5]
```diff
@@ -19,4 +20,12 @@ class TopLevelOrderRuleFixTest : FixTestBase("test/paragraph3/top_level", ::TopL
     fun `should fix top level order with comment`() {
         fixAndCompare("TopLevelWithCommentExpected.kt", "TopLevelWithCommentTest.kt")
     }
+
+    // FixMe:
+    @Disabled
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/TopLevelOrderRuleFixTest.kt:25]: Please add reason (`@Disabled("reason")`

**Review by petertrr — APPROVED** [review]

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/TopLevelOrderRuleFixTest.kt (hunk 5 above)
Comment (kgevorkyan) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/files/TopLevelOrderRuleFixTest.kt]: Done

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt (hunk 4 above)
Comment (kgevorkyan) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter1/PackageNaming.kt]: Thanks

**Review by kentr0w — APPROVED** [review]
lgtm

**Pull request merged**